src/PowerShellEditorServices/Session/PromptNest.cs

```diff
@@ -0,0 +1,564 @@
+                        _versionSpecificOperations.StopCommandInDebugger(_powerShellContext);
+                        currentFrame.ThreadController.StartThreadExit(DebuggerResumeAction.Stop);
+                        currentFrame.WaitForFrameExit(CancellationToken.None);
+                        continue;
+                    }
+
+                    if (currentFrame.FrameType.HasFlag(PromptNestFrameType.NestedPrompt))
+                    {
+                        _powerShellContext.ExitAllNestedPrompts();
+                        continue;
+                    }
+
+                    currentFrame.PowerShell.BeginStop(null, null);
+                    currentFrame.WaitForFrameExit(CancellationToken.None);
+                }
+
+                _consoleReader?.StopCommandLoop();
+                _readLineFrame.Dispose();
+                CurrentFrame.Dispose();
+                _frameStack.Clear();
+                _powerShellContext = null;
+                _consoleReader = null;
+                _isDisposed = true;
+            }
+        }
+
+        /// <summary>
+        /// Gets the <see cref="ThreadController" /> for the current frame.
+        /// </summary>
+        /// <returns>
+        /// The <see cref="ThreadController" /> for the current frame, or
+        /// <see langword="null" /> if the current frame does not have one.
+        /// </returns>
+        internal ThreadController GetThreadController()
+        {
+            if (_isDisposed)
+            {
+                return null;
+            }
+
+            return CurrentFrame.IsThreadController ? CurrentFrame.ThreadController : null;
+        }
+
+        /// <summary>
+        /// Create a new <see cref="PromptNestFrame" /> and set it as the current frame.
+        /// </summary>
+        internal void PushPromptContext()
+        {
+            if (_isDisposed)
+            {
+                return;
+            }
+
+            PushPromptContext(PromptNestFrameType.Normal);
+        }
+
+        /// <summary>
+        /// Create a new <see cref="PromptNestFrame" /> and set it as the current frame.
+        /// </summary>
+        /// <param name="frameType">The frame type.</param>
+        internal void PushPromptContext(PromptNestFrameType frameType)
+        {
+            if (_isDisposed)
+            {
+                return;
+            }
+
+            _frameStack.Push(
+                new PromptNestFrame(
+                    frameType.HasFlag(PromptNestFrameType.Remote)
+                        ? PowerShell.Create()
+                        : PowerShell.Create(RunspaceMode.CurrentRunspace),
+                    NewHandleQueue(),
+                    frameType));
+        }
+
+        /// <summary>
+        /// Dispose of the current <see cref="PromptNestFrame" /> and revert to the previous frame.
+        /// </summary>
+        internal void PopPromptContext()
+        {
+            PromptNestFrame currentFrame;
+            lock (_syncObject)
```

Comment:
what's the significance of this `_syncObject` and what does locking it do?

Reply:
Locking `_syncObject` is done to synchronize access to the frame stack. It prevents race conditions that would occur if two threads tried to pop a context at the same time.